[minor] overload from_unixtime func to have optional timezone parameter

buraksenn opened the pull request:
## Which issue does this PR close?
Not closes but discussed in #12892


## Rationale for this change
from_unixtime is not aware of timezone so this PR makes it have optional timezone parameter

## What changes are included in this PR?
code,test,doc change for overloaded from_unixtime

## Are these changes tested?
yes added tests

## Are there any user-facing changes?
yes

jonathanc-n reviewed (approved):
This LGTM! Thank you very much

Omega359 commented on docs/source/user-guide/sql/scalar_functions_new.md:
````diff
@@ -2003,12 +2003,13 @@ _Alias of [date_trunc](#date_trunc)._
 
 ```
-from_unixtime(expression)
+from_unixtime(expression, timezone)
````
The common way to denote optional parameters is with [, optional_param]
```suggestion
from_unixtime(expression[, timezone])
```

Omega359 commented on datafusion/functions/src/datetime/from_unixtime.rs:
```diff
@@ -93,12 +124,59 @@ fn get_from_unixtime_doc() -> &'static Documentation {
             .with_argument(
                 "expression",
-                "Expression to operate on. Can be a constant, column, or function, and any combination of arithmetic operators."
+                "Expression to operate on. Can be a constant, column, or function, and any combination of arithmetic operators.",
```
This probably should be just `.with_stardard_argument("expression", None)`

Omega359 commented on datafusion/functions/src/datetime/from_unixtime.rs:
```diff
@@ -93,12 +124,59 @@ fn get_from_unixtime_doc() -> &'static Documentation {
 }
+
+#[cfg(test)]
+mod test {
```
These tests are great to have but it would be very nice to have a few additional ones in timestamps.slt as well

buraksenn replied:
added

Omega359 commented on datafusion/functions/src/datetime/from_unixtime.rs:
```diff
@@ -93,12 +124,59 @@ fn get_from_unixtime_doc() -> &'static Documentation {
             .with_doc_section(DOC_SECTION_DATETIME)
             .with_description("Converts an integer to RFC3339 timestamp format (`YYYY-MM-DDT00:00:00.000000000Z`). Integers and unsigned integers are interpreted as nanoseconds since the unix epoch (`1970-01-01T00:00:00Z`) return the corresponding timestamp.")
-            .with_syntax_example("from_unixtime(expression)")
+            .with_syntax_example("from_unixtime(expression, timezone)")
```
A SQL example would be a very nice addition to have. The way I typically do that is to [build the CLI](https://datafusion.apache.org/user-guide/cli/index.html) with docker, run it and execute the desired sql in the cli and copy/paste it into the docs here.

alamb replied:
I actually did this and found a bug: 

```sql
> select from_unixtime(1730592000, 'America/New_York');
Internal error: Failed due to a difference in schemas, original schema: DFSchema { inner: Schema { fields: [Field { name: "from_unixtime(ts_data.ts / Int64(1000000000),Utf8(\"America/New_York\"))", data_type: Timestamp(Second, Some("Utf8")), nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "from_unixtime(Int64(1730592000),Utf8(\"America/New_York\"))", data_type: Timestamp(Second, Some("Utf8")), nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "arrow_typeof(from_unixtime(Int64(1730592000),Utf8(\"America/New_York\")))", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }], metadata: {} }, field_qualifiers: [None, None, None], functional_dependencies: FunctionalDependencies { deps: [] } }, new schema: DFSchema { inner: Schema { fields: [Field { name: "from_unixtime(ts_data.ts / Int64(1000000000),Utf8(\"America/New_York\"))", data_type: Timestamp(Second, Some("Utf8")), nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "from_unixtime(Int64(1730592000),Utf8(\"America/New_York\"))", data_type: Timestamp(Second, Some("America/New_York")), nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }, Field { name: "arrow_typeof(from_unixtime(Int64(1730592000),Utf8(\"America/New_York\")))", data_type: Utf8, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} }], metadata: {} }, field_qualifiers: [None, None, None], functional_dependencies: FunctionalDependencies { deps: [] } }.
```

alamb replied:
I think the difference is that you need to take the *value* of the timezone argument, not just the type
 
You can fix this by using `return_type_from_exprs` instead of `return_type`:

https://github.com/apache/datafusion/blob/b7f4db4b56e945944c0448f84c61b6f3b86728f8/datafusion/expr/src/udf.rs#L451

buraksenn replied:
Thanks, @alamb I could not find time before today I was debugging the issue. Fixed it with what you've said and added an SQL example

jonathanc-n commented:
I can probably finish the rest of the issue in another pr after this is merged

alamb commented:
I took the liberty of merging this PR up from main to resolve a conflict

alamb reviewed (commented):
Thanks @buraksenn -- I am marking this PR as draft as it is no longer waiting on review (and I am trying to clear the review queue)

I did merge again from main to fix the compile error

buraksenn commented:
> Thanks @buraksenn -- I am marking this PR as draft as it is no longer waiting on review (and I am trying to clear the review queue)
> 
> I did merge again from main to fix the compile error

Thanks for the review. I've fixed the issue and added test cases to timestamps.slt. I'm converting it to Ready for review again

berkaysynnada commented on datafusion/functions/src/datetime/from_unixtime.rs:
```diff
@@ -59,28 +63,63 @@ impl ScalarUDFImpl for FromUnixtimeFunc {
+
     fn return_type(&self, _arg_types: &[DataType]) -> Result<DataType> {
-        Ok(Timestamp(Second, None))
+        internal_err!("return_type should not be called")
```
Can you direct to `return_type_from_exprs`

buraksenn replied:
In the docs it says that if `return_type_from_expr` is implemented we should return internal error so that's why I did this.

https://github.com/buraksenn/datafusion/blob/b7f4db4b56e945944c0448f84c61b6f3b86728f8/datafusion/expr/src/udf.rs#L419-L421: 
```
    /// What [`DataType`] will be returned by this function, given the types of
    /// the arguments.
    ///
    /// # Notes
    ///
    /// If you provide an implementation for [`Self::return_type_from_exprs`],
    /// DataFusion will not call `return_type` (this function). In this case it
    /// is recommended to return [`DataFusionError::Internal`].
```

buraksenn replied:
aa got it. done

berkaysynnada reviewed (approved):
LGTM, thank you @buraksenn
